src/test/neorados/common_tests.h

```diff
@@ -15,27 +15,141 @@
 #include <string>
 #include <string_view>
 
+#include <boost/asio.hpp>
+#include <boost/asio/awaitable.hpp>
+
+#include <boost/system/error_code.hpp>
+
 #include "include/neorados/RADOS.hpp"
 
 std::string get_temp_pool_name(std::string_view prefix = {});
 
 template<typename CompletionToken>
-auto create_pool(neorados::RADOS& r, std::string pname,
-		 CompletionToken&& token)
+auto create_pool(
+  neorados::RADOS& r,
+  std::string pname,
+  CompletionToken&& token)
+{
+  namespace ba = boost::asio;
+  namespace bs = boost::system;
+  auto&& executor = ba::get_associated_executor(token);
+  std::cerr << "In create_pool" << std::endl;
+  return ba::async_compose<CompletionToken, void(bs::error_code, std::int64_t)>
+    (
+      [&executor, &r, &pname](auto& self) {
+	std::cerr << "In create_pool's async_compose" << std::endl;
+	co_spawn(
+	  executor,
+	  [&r, &self, pname]() mutable -> ba::awaitable<void> {
+	    std::cerr << "In create_pool's co_spawn" << std::endl;
+	    try {
+	      std::cerr << "calling r.create_pool" << std::endl;
```

Comment:
@cbodley Do you have an idea what I might be doing wrong here?

The main test function, for the moment, has everything `ifdef`'d out apart from creating the pool.

When I try running it, I get:
```
[ RUN      ] neocls_rgw.test_version_inc_read
This is some noise! 0
This is some noise! 1
Calling make_temp_pool.
In async_compose
Calling create_pool
In create_pool
In create_pool's async_compose
In create_pool's co_spawn
calling r.create_pool
[       OK ] neocls_rgw.test_version_inc_read (57 ms)
```

The pool actually *does* get created. So it looks like it's creating the pool, then falling out of the run loop without executing any more of the coroutine.

As I understand it, the `async_compose` in `make_temp_pool`should be keeping the queues in the main coroutine and `io_context` alive. And the `async_compose` in `create_pool` should be keeping the queues in `make_temp_pool`'s coroutine and `io_context` alive.

So I'm not sure what's causing it to get the wrong continuation.

Reply:
> As I understand it, the async_compose in make_temp_poolshould be keeping the queues in the main coroutine and io_context alive

`async_compose()` does track 'work' on the `executor, r.get_executor()` arguments you gave it to prevent their io_context(s) from stopping. but `async_compose()` is only calling `co_spawn()` then returning, so it doesn't cover the coroutine's entire execution

i would have assumed that `co_spawn(executor` would track 'work' on executor until the coroutine returns, but that doesn't seem to be the case - instead, we need to rely on each of its `co_await`s to track their work individually

here it looks like `co_await r.create_pool(pname, std::nullopt, ba::use_awaitable)` is the culprit, since `r.create_pool returned` is the first line we *don't* see in the output. however, the implementation of `create_pool()` looks right to me:
```c++
  SimpleOpComp::create(get_executor(), std::move(handler))
```
this completion should be tracking work both on `get_executor()` and `handler`'s associated executor. maybe you can find a way to step through the execution of just that `create_pool()` call?

Reply:
this might be a good place to use BOOST_ASIO_ENABLE_HANDLER_TRACKING: https://www.boost.org/doc/libs/1_79_0/doc/html/boost_asio/overview/core/handler_tracking.html